code/modules/chemistry/Reagents-Drugs.dm

```diff
@@ -289,53 +289,89 @@ datum
+				"The universe itself",
+			)
 
 			on_mob_life(var/mob/M, var/mult = 1)
 				if(!M) M = holder.my_atom
-				// TODO. Write awesome hallucination algorithm!
-//				if(M.canmove) step(M, pick(cardinal))
-//				if(prob(7)) M.emote(pick("twitch","drool","moan","giggle"))
-				if(M.client && prob(20))
+				src.counter += mult //around half realtime
+				if(M.client && counter >= 6 && prob(20))
 					if(src.current_color_pattern == 1)
 						animate_fade_drug_inbetween_1(M.client, 40)
 						src.current_color_pattern = 2
 					else
 						animate_fade_drug_inbetween_2(M.client, 40)
 						src.current_color_pattern = 1
-				if(probmult(6))
-					switch(rand(1,2))
-						if(1)
-							if(prob(50))
-								fake_attack(M)
-							else
-								var/monkeys = rand(1,3)
-								for(var/i = 0, i < monkeys, i++)
-									fake_attackEx(M, 'icons/mob/monkey.dmi', "monkey_hallucination", "monkey ([rand(1, 1000)])")
-						if(2)
-							var/halluc_state = null
-							var/halluc_name = null
-							switch(rand(1,5))
-								if(1)
-									halluc_state = "pig"
-									halluc_name = pick("pig", "DAT FUKKEN PIG")
-								if(2)
-									halluc_state = "spider"
-									halluc_name = pick("giant black widow", "aw look a spider", "OH FUCK A SPIDER")
-								if(3)
-									halluc_state = "dragon"
-									halluc_name = pick("dragon", "Lord Cinderbottom", "SOME FUKKEN LIZARD THAT BREATHES FIRE")
-								if(4)
-									halluc_state = "slime"
-									halluc_name = pick("red slime", "some gooey thing", "ANGRY CRIMSON POO")
-								if(5)
-									halluc_state = "shambler"
-									halluc_name = pick("shambler", "strange creature", "OH GOD WHAT THE FUCK IS THAT THING?")
-							fake_attackEx(M, 'icons/effects/hallucinations.dmi', halluc_state, halluc_name)
-				if(probmult(9))
-					M.playsound_local(M.loc, pick("explosion", "punch", 'sound/vox/poo-vox.ogg', "clownstep", 'sound/weapons/armbomb.ogg', 'sound/weapons/Gunshot.ogg'), 50, 1)
+				if(probmult(12) && !ON_COOLDOWN(M, "hallucination_spawn", 30 SECONDS))
+					if (prob(30))
+						if(prob(50))
+							fake_attack(M)
+						else
+							var/monkeys = rand(1,3)
+							for(var/i = 0, i < monkeys, i++)
+								fake_attackEx(M, 'icons/mob/monkey.dmi', "monkey_hallucination", pick_string_autokey("names/monkey.txt"))
+					else
+						var/fake_type = pick(childrentypesof(/obj/fake_attacker))
+						new fake_type(M.loc, M)
+				if(probmult(min(16 + src.counter/2, 30))) //THE VOICES GET LOUDER
+					var/atom/origin = M.loc
+					var/turf/mob_turf = get_turf(M)
+					if (mob_turf)
+						origin = locate(mob_turf.x + rand(-10,10), mob_turf.y + rand(-10,10), mob_turf.z)
+					//wacky loosely typed code ahead
+					var/datum/hallucinated_sound/chosen = pick(src.halluc_sounds)
+					if (istype(chosen)) //it's a datum
+						SPAWN(0)
+							for (var/i = 1 to rand(chosen.min_count, chosen.max_count))
+								M.playsound_local(origin, chosen.path, 100, 1, pitch = chosen.pitch)
+								sleep(chosen.delay)
```

Comment:
wouldn't it make more sense for this to be a proc on the `hallucinated_sound` datum?